runtime/wasm/src/host.rs

```diff
@@ -51,41 +43,33 @@ where
     }
 }
 
-impl<T, L, S> RuntimeHostBuilder<T, L, S>
+impl<S> RuntimeHostBuilder<S>
 where
-    T: EthereumAdapter,
-    L: LinkResolver,
-    S: Store,
+    S: Store + SubgraphDeploymentStore + EthereumCallCache,
 {
     pub fn new(
-        ethereum_adapters: HashMap<String, Arc<T>>,
-        link_resolver: Arc<L>,
+        ethereum_adapters: HashMap<String, Arc<impl EthereumAdapter>>,
+        link_resolver: Arc<dyn LinkResolver>,
         stores: HashMap<String, Arc<S>>,
     ) -> Self {
         RuntimeHostBuilder {
-            ethereum_adapters,
+            ethereum_adapters: ethereum_adapters
```

Comment:
So I've now resolved this by using `dyn EthereumAdapter` when this is first created. `impl` in an argument is just sugar for a generic, `dyn` is a trait object. I'm using trait objects here simply because they are syntatically less noisy and avoid needing a generic parameter in structs.

The reason why that cast is more than just changing the types is because `Arc<impl EthereumAdapter>` and `Arc<dyn EthereumAdapter>` have different representations, the second one is a fat pointer, so it's actually two pointers, one for the vtable an another for the object. The cast is adding that second pointer. The really technical reason is that Arc is [CoercedUnsized](https://doc.rust-lang.org/std/ops/trait.CoerceUnsized.html) while `HashMap` isn't because it doesn't have direct access to the pointers.